### PR TITLE
Emit Run 0 SARIF for accepted Type-1 and Type-2 clone pairs

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Implements Run 0 emission for accepted Type-1 and Type-2 clone pairs. Adds a new run0 module under `crates/whitaker_clones_core/` with `accept_candidate_pairs` and `emit_run0`, plus unit tests and BDD coverage, and updates docs and roadmap accordingly.

## What’s included
- New module subtree: `crates/whitaker_clones_core/src/run0/` containing:
  - `emit.rs` (SARIF emission logic)
  - `error.rs` (typed Run 0 errors)
  - `mod.rs` (module wiring)
  - `score.rs` (integer-based Jaccard scoring and thresholds)
  - `span.rs` (byte-range to SARIF region conversion)
  - `types.rs` (public input/output types)
  - `tests.rs` (unit tests for Run 0 logic)
- Public API exposure in `crates/whitaker_clones_core/src/lib.rs` for Run 0:
  - `AcceptedPair`, `Run0Error`, `Run0Result`, `SimilarityRatio`, `SimilarityThreshold`, `TokenFragment`, `TokenPassConfig`, `accept_candidate_pairs`, `emit_run0`.
- Behavioural and integration tests:
  - `crates/whitaker_clones_core/src/run0_sarif.feature` (BDD feature file)
  - `crates/whitaker_clones_core/src/run0_sarif_behaviour.rs` (BDD harness)
- Tests and fixtures for Run 0 emission:
  - `crates/whitaker_clones_core/tests/features/run0_sarif.feature` (feature area)
  - `crates/whitaker_clones_core/tests/run0_sarif_behaviour.rs` (behaviour tests)
- ExecPlan living document updated to reflect implementation:
  - `docs/execplans/7-2-3-emit-sarif-run-0-for-accepted-type-1-and-type-2-pairs.md` (status COMPLETE)
- Roadmap and design updates:
  - `docs/roadmap.md` updated to mark 7.2.3 as done
  - `docs/whitaker-clone-detector-design.md` updated with dedicated 7.2.3 decisions (in-file section)

## Rationale
- This completes roadmap item 7.2.3 by turning the prior planning into a concrete, test-covered Run 0 emission path. The new `run0` module reuses existing SARIF tooling (`whitaker_sarif`) and keeps emission logic encapsulated within `whitaker_clones_core` as planned, enabling deterministic results and stable regions for downstream tooling.

## Implementation plan (high level)
- Stage A: Draft ExecPlan (completed)
- Stage B: Add unit tests for thresholds, span mapping, and deterministic ordering (completed)
- Stage C: Add BDD coverage for Run 0 emission (completed)
- Stage D: Implement run0 module skeleton and API surface (completed)
- Stage E: Implement Run 0 emission logic (completed)
- Stage F: Documentation updates (`## Implementation decisions (7.2.3)`) (completed)
- Stage G: Mark roadmap item done in `docs/roadmap.md` (completed)
- Stage H: Validate with all gates (fmt, lint, test, etc.) (completed)
- Stage I: Finalize ExecPlan post-implementation (completed)

## How to review
- Verify the new Run 0 API surface and the emitted SARIF run shape align with the design goals (one SARIF result per accepted pair, with stable primary/related locations and fingerprints).
- Confirm deterministic sorting and deduplication behavior for multiple accepted pairs.
- Check unit and behavioural tests exercise threshold boundaries, multi-line span conversion, and reversed-pair determinism.
- Review documentation updates for accuracy and consistency with the code changes.

## Documentation impact
- Introduces a new ExecPlan living document updated to COMPLETE status.
- Roadmap and design docs reflect the implemented Run 0 emission decisions and outcomes.

## Acceptance criteria
- The PR implements Run 0 emission for Type-1 and Type-2 accepted pairs with deterministic results.
- Unit tests and BDD tests cover happy paths, threshold boundaries, multi-line spans, and reversed inputs.
- The emitted SARIF run uses the existing `whitaker_sarif` tooling and carries stable fingerprints and 1-based SARIF regions.
- Documentation and roadmap reflect completion of 7.2.3.

📎 **Task**: https://www.devboxer.com/task/81bbb2d8-2581-49d3-a21b-cba40d1d3c53

📎 **Task**: https://www.devboxer.com/task/3515b49a-8747-49b6-80b3-5e2575dce6dd

## Summary by Sourcery

Implement token-pass Run 0 SARIF emission for accepted clone pairs and expose a public API for acceptance and emission in whitaker_clones_core.

New Features:
- Add a dedicated run0 module in whitaker_clones_core for token-pass acceptance and SARIF Run 0 emission.
- Expose public types and functions for Run 0, including TokenFragment, TokenPassConfig, AcceptedPair, SimilarityThreshold, accept_candidate_pairs, and emit_run0.
- Introduce deterministic Jaccard-based scoring and threshold handling for Type-1 and Type-2 clone pairs.

Enhancements:
- Define stable pair and token hash fingerprints using SHA-256 for SARIF partialFingerprints.
- Implement byte-range to SARIF region mapping with validation for UTF-8 alignment and bounds.
- Ensure deterministic sorting and deduplication of emitted SARIF results based on rule, fingerprints, and message.

Documentation:
- Document implementation decisions for roadmap item 7.2.3 in the clone-detector design and add a detailed ExecPlan for Run 0 emission.
- Update the roadmap to mark SARIF Run 0 emission for accepted Type-1 and Type-2 pairs as complete.

Tests:
- Add unit tests for similarity scoring, threshold boundaries, span mapping, acceptance logic, and Run 0 emission behaviour.
- Add BDD feature scenarios and harness for end-to-end Run 0 SARIF emission, including edge cases for empty inputs, multi-line spans, and reversed pairs.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

## Code Duplication
crates/whitaker_clones_core/src/run0/tests.rs: 

### What lead to degradation?
The module contains 5 functions with similar structure: boundary_threshold_accepts_type1_pair,boundary_threshold_accepts_type2_pair,just_below_threshold_is_rejected,multi_line_region_tracks_trailing_newline and 1 more functions

### Why does this problem occur?
Duplicated code often leads to code that's harder to change since the same logical change has to be done in multiple functions. More duplication gives lower code health.

### How to fix it?
A certain degree of duplicated code might be acceptable. The problems start when it is the same behavior that is duplicated across the functions in the module, ie. a violation of the Don't Repeat Yourself (DRY) principle. DRY violations lead to code that is changed together in predictable patterns, which is both expensive and risky. DRY violations can be identified using CodeScene's X-Ray analysis to detect clusters of change coupled functions with high code similarity. [Read More](https://codescene.com/blog/software-revolution-part3/)
Once you have identified the similarities across functions, look to extract and encapsulate the concept that varies into its own function(s). These shared abstractions can then be re-used, which minimizes the amount of duplication and simplifies change.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

## Code Duplication
crates/whitaker_clones_core/src/run0/tests.rs: 

### What lead to degradation?
The module contains 5 functions with similar structure: boundary_threshold_accepts_type1_pair,boundary_threshold_accepts_type2_pair,just_below_threshold_is_rejected,multi_line_region_tracks_trailing_newline and 1 more functions

### Why does this problem occur?
Duplicated code often leads to code that's harder to change since the same logical change has to be done in multiple functions. More duplication gives lower code health.

### How to fix it?
A certain degree of duplicated code might be acceptable. The problems start when it is the same behavior that is duplicated across the functions in the module, ie. a violation of the Don't Repeat Yourself (DRY) principle. DRY violations lead to code that is changed together in predictable patterns, which is both expensive and risky. DRY violations can be identified using CodeScene's X-Ray analysis to detect clusters of change coupled functions with high code similarity. [Read More](https://codescene.com/blog/software-revolution-part3/)
Once you have identified the similarities across functions, look to extract and encapsulate the concept that varies into its own function(s). These shared abstractions can then be re-used, which minimizes the amount of duplication and simplifies change.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

Annul any requirements that violate the en-GB-oxendict spelling (-ize / -yse / -our) conventions (for example a request to replace "normalize" with "normalise" or "artefact" with "artifact"), or where the requirement unnecessarily increases cyclomatic complexity.

~~~
Please address the comments from this code review:

## Individual Comments

### Comment 1
<location path="crates/whitaker_clones_core/src/run0/score.rs" line_range="159-171" />
<code_context>
+        >= score.union.saturating_mul(threshold.numerator)
+}
+
+fn repeated_division(numerator: usize, denominator: usize) -> Option<(usize, usize)> {
+    if denominator == 0 {
+        return None;
+    }
+
+    let mut quotient = 0usize;
+    let mut remainder = numerator;
+    while remainder >= denominator {
+        remainder = remainder.saturating_sub(denominator);
+        quotient = quotient.saturating_add(1);
+    }
+    Some((quotient, remainder))
+}
</code_context>
<issue_to_address>
**suggestion:** Simplify division to use native operators now that unions are non-zero

Given unions are guaranteed non‑zero at all current call sites, the helper can be replaced with direct `numerator / denominator` and `numerator % denominator`, which is clearer and avoids the manual loop. If you prefer to keep the helper, consider adding a debug assertion that `denominator > 0` to document the invariant.

```suggestion
fn repeated_division(numerator: usize, denominator: usize) -> Option<(usize, usize)> {
    if denominator == 0 {
        debug_assert!(denominator != 0, "denominator must be non-zero");
        return None;
    }

    let quotient = numerator / denominator;
    let remainder = numerator % denominator;

    Some((quotient, remainder))
}
```
</issue_to_address>

### Comment 2
<location path="crates/whitaker_clones_core/src/run0/tests.rs" line_range="305-325" />
<code_context>
+}
+
+#[test]
+fn invalid_range_produces_typed_error() {
+    let error = region_for_range("alpha", "fn a() {}\n", 9..12)
+        .err()
+        .unwrap_or_else(|| panic!("invalid range must error"));
+
+    match error {
+        Run0Error::InvalidFingerprintRange {
+            fragment_id,
+            start,
+            end,
+            source_len,
+        } => {
+            assert_eq!(fragment_id, "alpha");
+            assert_eq!(start, 9);
+            assert_eq!(end, 12);
+            assert_eq!(source_len, 10);
+        }
+        other => panic!("unexpected error: {other}"),
+    }
+}
</code_context>
<issue_to_address>
**suggestion (testing):** Add a unit test for `InvalidUtf8Boundary` to complement the `InvalidFingerprintRange` coverage

`span::region_for_range` can also return `Run0Error::InvalidUtf8Boundary`, but the tests currently only cover `InvalidFingerprintRange` (out-of-bounds range). To cover the UTF-8 boundary path, please add a test along these lines:

```rust
#[test]
fn invalid_utf8_boundary_produces_typed_error() {
    // "á" = 2 bytes in UTF-8
    let source = "aáb";
    let mid = 1; // inside "á" byte sequence
    let err = region_for_range("alpha", source, 0..mid)
        .err()
        .unwrap_or_else(|| panic!("invalid utf-8 boundary must error"));

    match err {
        Run0Error::InvalidUtf8Boundary { fragment_id } => {
            assert_eq!(fragment_id, "alpha");
        }
        other => panic!("unexpected error: {other}"),
    }
}
```

This ensures we lock in the intended behaviour for non-ASCII sources and span computation around UTF-8 boundaries.

```suggestion
#[test]
fn invalid_range_produces_typed_error() {
    let error = region_for_range("alpha", "fn a() {}\n", 9..12)
        .err()
        .unwrap_or_else(|| panic!("invalid range must error"));

    match error {
        Run0Error::InvalidFingerprintRange {
            fragment_id,
            start,
            end,
            source_len,
        } => {
            assert_eq!(fragment_id, "alpha");
            assert_eq!(start, 9);
            assert_eq!(end, 12);
            assert_eq!(source_len, 10);
        }
        other => panic!("unexpected error: {other}"),
    }
}

#[test]
fn invalid_utf8_boundary_produces_typed_error() {
    // "á" = 2 bytes in UTF-8; index 2 is in the middle of that code point
    let source = "aáb";
    let mid = 2; // inside "á" byte sequence
    let error = region_for_range("alpha", source, 0..mid)
        .err()
        .unwrap_or_else(|| panic!("invalid utf-8 boundary must error"));

    match error {
        Run0Error::InvalidUtf8Boundary { fragment_id } => {
            assert_eq!(fragment_id, "alpha");
        }
        other => panic!("unexpected error: {other}"),
    }
}
```
</issue_to_address>

### Comment 3
<location path="crates/whitaker_clones_core/src/run0/tests.rs" line_range="153-162" />
<code_context>
+}
+
+#[test]
+fn duplicate_hashes_do_not_inflate_jaccard_score() {
+    let left = [
+        fingerprint(1, 0..3),
+        fingerprint(1, 3..6),
+        fingerprint(2, 6..9),
+    ];
+    let right = [fingerprint(1, 0..3), fingerprint(2, 3..6)];
+
+    let score = jaccard_similarity(&left, &right)
+        .unwrap_or_else(|| panic!("score should be present for non-empty fragments"));
+
+    assert_eq!(score, SimilarityRatio::new(2, 2));
+}
+
</code_context>
<issue_to_address>
**suggestion (testing):** Also test `jaccard_similarity` behaviour when one side has no fingerprints

The helper returns `None` when either side has no unique hashes, but current tests only cover non-empty inputs and deduplication. Please add a test like:

```rust
#[test]
fn jaccard_returns_none_for_empty_fragments() {
    let non_empty = [fingerprint(1, 0..3)];
    let empty: [Fingerprint; 0] = [];

    assert!(jaccard_similarity(&empty, &non_empty).is_none());
    assert!(jaccard_similarity(&non_empty, &empty).is_none());
}
```

so the `None`-on-empty contract is explicitly covered.

Suggested implementation:

```rust
}
+
+#[test]
+fn duplicate_hashes_do_not_inflate_jaccard_score() {
+    let left = [
+        fingerprint(1, 0..3),
+        fingerprint(1, 3..6),
+        fingerprint(2, 6..9),
+    ];
+    let right = [fingerprint(1, 0..3), fingerprint(2, 3..6)];
+
+    let score = jaccard_similarity(&left, &right)
+        .unwrap_or_else(|| panic!("score should be present for non-empty fragments"));
+
+    assert_eq!(score, SimilarityRatio::new(2, 2));
+}
+
+#[test]
+fn jaccard_returns_none_for_empty_fragments() {
+    let non_empty = [fingerprint(1, 0..3)];
+    let empty: [Fingerprint; 0] = [];
+
+    assert!(jaccard_similarity(&empty, &non_empty).is_none());
+    assert!(jaccard_similarity(&non_empty, &empty).is_none());
+}
+
+#[test]

```

This change assumes:
1. The `fingerprint` helper and `Fingerprint` type are already defined and in scope in this test module (they are referenced by existing tests).
2. `jaccard_similarity` is available from the imported `score` module, which is already used by `duplicate_hashes_do_not_inflate_jaccard_score`.

If the stray `#[test]` before the imports in your snippet is not actually present in the real file, adjust the SEARCH/REPLACE context to match the real surrounding lines (keep the two test functions adjacent and before the next item in the module).
</issue_to_address>

### Comment 4
<location path="crates/whitaker_clones_core/src/run0/tests.rs" line_range="76-85" />
<code_context>
+    );
+}
+
+#[test]
+fn emit_run0_sorts_and_deduplicates_results() {
+    let fragments = vec![
+        fragment(FragmentInput {
</code_context>
<issue_to_address>
**suggestion (testing):** Add unit tests for `MissingFragment` and `MixedProfiles` error paths in acceptance/emission

The current tests for `accept_candidate_pairs` and `emit_run0` cover happy paths and boundaries but don’t exercise the `Run0Error::MissingFragment` or `Run0Error::MixedProfiles` variants.

Please add tests that:
- Call `accept_candidate_pairs` with a candidate referencing a non-existent fragment and assert you get `Run0Error::MissingFragment { fragment_id: "beta" }`.
- Call `accept_candidate_pairs` with fragments using different `NormProfile`s and assert you get `Run0Error::MixedProfiles { left_fragment: "alpha", right_fragment: "beta" }`.

This will ensure the acceptance-stage error modes are covered and remain stable over time.
</issue_to_address>

### Comment 5
<location path="docs/execplans/7-2-3-emit-sarif-run-0-for-accepted-type-1-and-type-2-pairs.md" line_range="28-29" />
<code_context>
+   pass.
+2. Accepted Type-1 pairs produce `WHK001` results and accepted Type-2 pairs
+   produce `WHK002` results.
+3. Each emitted result carries deterministic `message`, `locations`,
+   `related_locations`, `partial_fingerprints`, and Whitaker `properties`.
+4. Unit tests cover happy paths, unhappy paths, threshold boundaries,
+   multi-line span conversion, stable ordering, and duplicate suppression.
</code_context>
<issue_to_address>
**suggestion (typo):** Use `relatedLocations` here for consistency with SARIF naming and the rest of the docs.

Earlier in the design docs this field is called `relatedLocations` / `relatedLocation`, matching the SARIF spec. Using `related_locations` here (in backticks with other SARIF field names) looks like an inconsistent typo rather than an intentional Rust identifier. Please align it to `relatedLocations` for consistency.

```suggestion
3. Each emitted result carries deterministic `message`, `locations`,
   `relatedLocations`, `partial_fingerprints`, and Whitaker `properties`.
```
</issue_to_address>

### Comment 6
<location path="docs/execplans/7-2-3-emit-sarif-run-0-for-accepted-type-1-and-type-2-pairs.md" line_range="1" />
<code_context>
+# Emit SARIF Run 0 for accepted Type-1 and Type-2 clone pairs (roadmap 7.2.3)
+
+This ExecPlan is a living document. The sections `Constraints`, `Tolerances`,
</code_context>
<issue_to_address>
**suggestion (review_instructions):** Define SARIF on first use so the acronym is clear to readers unfamiliar with it.

Please expand SARIF on first use, for example: `Emit Static Analysis Results Interchange Format (SARIF) Run 0 ...`, and then use `SARIF` thereafter.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>

### Comment 7
<location path="docs/execplans/7-2-3-emit-sarif-run-0-for-accepted-type-1-and-type-2-pairs.md" line_range="44" />
<code_context>
+
+## Constraints
+
+- Scope only roadmap item 7.2.3. Do not implement filesystem discovery,
+  grouping into clone classes, abstract syntax tree refinement, HTML output, or
+  CLI wiring in this change.
</code_context>
<issue_to_address>
**suggestion (review_instructions):** CLI should be expanded on first use to satisfy the acronym-definition requirement.

In this section, `CLI wiring` appears without defining CLI. Consider wording such as `command-line interface (CLI) wiring` at first mention, and then `CLI` thereafter.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>
~~~

File: crates/whitaker_clones_core/src/run0/span.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    }
+    starts
+}
+
+fn line_and_column(starts: &[usize], offset: usize) -> (usize, usize) {
+    let line_index = starts
+        .partition_point(|start| *start <= offset)
+        .saturating_sub(1);
+    let line_start = starts[line_index];
+    let column = offset.saturating_sub(line_start).saturating_add(1);
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/whitaker_clones_core/src/run0/span.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: crates/whitaker_clones_core/src/run0/tests.rs [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+    assert_eq!(
+        (
+            extracted.profile.as_str(),
+            extracted.k,
+            extracted.window,
+            extracted.class_size,
+        ),
+        ("T2", 25, 16, 2),
+    );
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/whitaker_clones_core/src/run0/tests.rs:350]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: crates/whitaker_clones_core/src/run0/tests_emit.rs [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+            .iter()
+            .map(|(hash, range)| fingerprint(*hash, range.clone()))
+            .collect(),
+    )
+}
+
+fn pair(left: &str, right: &str) -> crate::CandidatePair {
+    crate::CandidatePair::new(FragmentId::from(left), FragmentId::from(right))
+        .expect("pair `{left}` and `{right}` must be distinct")
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/whitaker_clones_core/src/run0/tests_emit.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai review
